ActiveStorageで画像アップロードを実装するの提出

YuSa0-6 opened the pull request:
branch名変更により自動で前回の[プルリクエスト](https://github.com/YuSa0-6/fjord-books_app-2023/pull/4)がCloseしたので新規作成します。

YuSa0-6 committed:
画像を扱うためにActiveStorageを入れる

YuSa0-6 committed:
各ユーザーに画像を添付するためにモデルを定義する

YuSa0-6 committed:
avatarのストロングパラメーターを設定する

YuSa0-6 committed:
アップロード用のフォームを追加

YuSa0-6 committed:
一覧ページにて画像がある場合は表示する

YuSa0-6 committed:
ユーザー詳細画面でのアイコン画像がある場合のみ表示する

YuSa0-6 committed:
大きすぎる画像の時、リサイズするように設定

YuSa0-6 committed:
mini_magickを導入してresizeするようにした。index,showに関しては事前にvariantを使用しているので元に戻した。

YuSa0-6 committed:
editページでアイコンを編集できるようにフォームを追加した

YuSa0-6 committed:
結局使用していなかったので削除します。目的としてはリサイズ用の設定を事前に設定できることでしたが、使用しませんでした。

YuSa0-6 committed:
mini_magickからvipsに戻す。

YuSa0-6 committed:
修正時に一箇所で済むように、パーシャル内にユーザーアイコン画像を設置する。

JunichiIto commented on Gemfile:
```diff
@@ -86,3 +86,4 @@ gem 'carrierwave'
 gem 'devise'
 gem 'devise-i18n'
 gem 'kaminari'
+gem 'mini_magick'
```
この記述が本当に必要かどうか、最新のRailsガイドを読んで確認してみてください〜。

JunichiIto commented on app/views/users/_user.html.erb:
```diff
@@ -1,5 +1,8 @@
 <div id="<%= dom_id user %>">
 
+  <p>
```
avatarがないときは空のpタグが出力されますね

JunichiIto commented on app/views/users/_user.html.erb:
```diff
@@ -1,5 +1,8 @@
 <div id="<%= dom_id user %>">
 
+  <p>
+    <%= image_tag user.avatar.variant(resize_to_limit: [400, 400]).processed if user.avatar.attached? %>
```
```suggestion
    <%= image_tag user.avatar.variant(resize_to_limit: [400, 400])if user.avatar.attached? %>
```

processed無しでも動きませんか？

YuSa0-6 replied:
動くと思います。

JunichiIto commented on app/views/devise/registrations/new.html.erb:
```diff
@@ -23,6 +23,11 @@
   </div>
 
+  <div>
+    <%= f.label :avatar %>
```
翻訳が必要では？

> 画像アップロード用に追加した項目はi18nのYAMLファイルを編集して項目名を表示させよう（英語化は考慮不要）

JunichiIto replied:
あれ、翻訳してない？
ブラウザ上では英語で"Avatar"のように出てるかもしれませんが、HTMLソースを見ると、"translation missing"みたいなエラーメッセージがHTMLコメントとして出力されているんじゃないかと思います

JunichiIto reviewed (commented):
コメントしました！
あと、「一覧画面に画像を表示させる場合はN+1問題を起こさないようにしよう」の考慮もお願いします〜。

YuSa0-6 committed:
mini_magickではなくvipsに変更するためにmini_magickをアンインストールする。
関係するカラムの方をbigint->intgerに変更になっている。

YuSa0-6 committed:
vipsに変更するのでmini_magickを削除しておく。

YuSa0-6 committed:
画像が表示されない場合にpタグのみ出力されてしまうことを回避するために<%= if %>にて分岐させた。

YuSa0-6 committed:
参考https://techracho.bpsinc.jp/hachi8833/2018_02_06/52179
N+1問題に対応する。

YuSa0-6 committed:
ユーザー登録における画像アップロードの翻訳

YuSa0-6 committed:
Revert "ユーザー登録における画像アップロードの翻訳"

This reverts commit 0fbb215c632ed292e33bfd318ee24d302760e3b9.

YuSa0-6 committed:
修正、クエリーが複数走っていたので対応できていないと判断ししたので一部記述を変更する。

JunichiIto commented on app/models/user.rb:
```diff
@@ -5,4 +5,5 @@ class User < ApplicationRecord
          :recoverable, :rememberable, :validatable
 
   has_one_attached :avatar
+  scope :with_eager_loaded_avater, -> { eager_load(avatar_attachment: :avatar) }
```
自分で定義しなくてもRailsが標準で用意してくれているのでそのscopeを使ってください〜。

JunichiIto replied:
不要になったscopeは削除して下さい〜。

JunichiIto commented on Gemfile:
```diff
@@ -86,4 +86,3 @@ gem 'carrierwave'
 gem 'devise'
 gem 'devise-i18n'
 gem 'kaminari'
-gem 'mini_magick'
```
今回はcarrierwaveがimage_processingをインストールしなくてもこのままでも動きますが、ActiveStorageはimage_processingが必要なので、該当箇所のコメントを外しておくのが安心です。

> 画像分析や画像加工のためにimage_processing gemも必要です。Gemfileのimage_processing gemをコメント解除するか、必要に応じて追加します。
https://railsguides.jp/active_storage_overview.html

JunichiIto commented on config/application.rb:
```diff
@@ -18,5 +18,6 @@ class Application < Rails::Application
     #
     # config.time_zone = "Central Time (US & Canada)"
     # config.eager_load_paths << Rails.root.join("extras")
+    config.active_storage.variant_processor = :vips
```
参考：デフォルトのプロセッサがvipsなので必ずしもこの設定は書かなくてもOKです。
（明示的に書いておく、という考え方もあるので今回はこのままでOK）

JunichiIto reviewed (commented):
コメントしました〜。

YuSa0-6 committed:
image_processingが必要なのでコメントを外しておく。

YuSa0-6 committed:
N+1対策のために追記する。参考https://edgeapi.rubyonrails.org/classes/ActiveStorage/Attached/Model.html

YuSa0-6 committed:
入力フォーム日本語対応。

YuSa0-6 committed:
Revert "N+1対策のために追記する。参考https://edgeapi.rubyonrails.org/classes/ActiveStorage/Attached/Model.html"

This reverts commit 923080d5d530cc43cc0db0be770308062f752f12.

JunichiIto commented on app/controllers/users_controller.rb:
```diff
@@ -2,7 +2,7 @@
 class UsersController < ApplicationController
   def index
-    @users = User.order(:id).page(params[:page])
+    @users = User.order(:id).page(params[:page]).with_attached_avatar
```
```suggestion
    @users = User.with_attached_avatar.order(:id).page(params[:page])
```

慣習的にはorderやpageは最後の方に書くことが多いので、こっちの方がいいかも

JunichiIto reviewed (commented):
コメントしました〜。

YuSa0-6 committed:
https://github.com/YuSa0-6/fjord-books_app-2023/pull/7#discussion_r1602329683
よりorderやpageは最後の方に書くことが多いため順番を入れ替える。

YuSa0-6 committed:
downしたmigrationをupする

YuSa0-6 committed:
いらないscopeは削除する。
review https://github.com/YuSa0-6/fjord-books_app-2023/pull/7#discussion_r1602329822

JunichiIto reviewed (approved):
確認しました！OKです〜🙆‍♂️
